Replace some Lodash function usages with vanilla equivalents

nwalters512 opened the pull request:
I'm lifting this out of #11024 to minimize the diff there; I applied the same changes to as many other places as I could.

reteps commented on apps/prairielearn/src/lib/load.ts:
```diff
@@ -42,7 +42,7 @@ class LoadEstimator {
     debug(`LoadEstimator.startJob(): jobType = ${this.jobType}, id = ${id}`);
     this._addIntegratedLoad();
-    if (_.has(this.currentJobs, id)) {
+    if (Object.prototype.hasOwnProperty.call(this.currentJobs, id)) {
```
Since currentJobs is just a record, would `id in this.currentJobs` be easier to read? Also in other places.

nwalters512 replied:
`id` could (in theory) itself be something like `"hasOwnProperty"`. I'm fine using `in` directly when the property is a static string. If it's dynamic, as it is here, I used the method you see here.

reteps commented on apps/prairielearn/src/question-servers/freeform.js:
```diff
@@ -448,7 +450,9 @@ function checkData(data, origData, phase) {
     if (!editPhases.includes(phase)) {
-      if (!_.has(origData, prop)) return '"' + prop + '" is missing from "origData"';
+      if (!Object.prototype.hasOwnProperty.call(origData, prop)) {
+        return '"' + prop + '" is missing from "origData"';
```
Nit: if we are cleaning this up, maybe change to a template literal?

nwalters512 replied:
I can do that, sure.

nwalters512 replied:
Done in 0ee9722.

reteps commented on apps/prairielearn/src/sync/course-db.ts:
```diff
@@ -266,6 +266,8 @@ export interface AssessmentAllowAccess {
   active: boolean;
   timeLimitMin: number;
   password: string;
+  showClosedAssessment: boolean;
```
Was this noticed when you removed some of the lodash mentions?

nwalters512 replied:
Yep - for `_.get(foo, 'bar')`, TypeScript won't check that `foo` has a `bar property, but naturally it will when you just do `foo.bar`.

reteps reviewed (approved):
I actually was thinking of doing some of these changes myself!

nwalters512 commented:
@reteps I fixed a bug that the tests caught in 5dbbc54. I also replaced `_.cloneDeep` with `structuredClone` in a few places. It'd be great if you could take another final look.

reteps commented:
> @reteps I fixed a bug that the tests caught in [5dbbc54](https://github.com/PrairieLearn/PrairieLearn/commit/5dbbc54a9918e2c706400c133dbcb6ce8b7e6fd0). I also replaced `_.cloneDeep` with `structuredClone` in a few places. It'd be great if you could take another final look.

The changes look good. So the hasownproperty is more restrictive than the `in` operator, TIL.

nwalters512 commented on apps/prairielearn/src/question-servers/freeform.js:
```diff
@@ -1461,30 +1466,30 @@ export async function render(
-              extensions[elementName][extensionName],
-            ).dynamicDependencies;
-            if (_.has(extension, 'extensionStyles')) {
+            const extension =
+              structuredClone(extensions[elementName][extensionName]).dependencies ?? {};
+            const extensionDynamic =
+              structuredClone(extensions[elementName][extensionName]).dynamicDependencies ?? {};
```
Makes sense! 👍

jonatanschroeder commented:
For consideration: https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore/blob/master/configuring.md :-)
